### ISSUE TITLE
[BUG] The created folders for the different users miss +x permissions

### ISSUE THREAD

**Issue by MohammedNoureldin** [docker-mailserver/docker-mailserver#2238]
### Miscellaneous first checks

- [X] I checked that all ports are open and not blocked by my ISP / hosting provider.
- [X] I know that SSL errors are likely the result of a wrong setup on the user side and not caused by DMS itself. I'm confident my setup is correct.

### Affected Component(s)

Mail home folder permissions (tested with v10.1.2 and 10.2.0)

### What happened and when does this occur?

```Markdown
When trying to use the mailserver, some permission error related to mail home directory is shown.
```

### What did you expect to happen?

```Markdown
To not get the error
```

### How do we replicate the issue?

```Markdown
Try the attached configuration with LDAP server.
```

### DMS version

Tested with 10.1.2 and 10.2.0

### How much RAM is available to DMS explicitly?

less than 2GB

### How many CPU cores are available?

less than 1 Core

### Is DMS running in a virtualized environment?

... a virtual private server (VPS) (with virtual CPU cores)

### What operating system is DMS running on?

Linux

### What instruction set architecture is DMS running on?

x86_64 / AMD64

### I/O - Persistent memory

-

### What container orchestration tool are you using?

Docker Compose

### Docker version

Docker version 20.10.8, build 3967b7d

### Docker Compose version

docker-compose version 1.29.2, build 5becea4c

### The output of `uname -a`

Linux srv02 5.4.0-84-generic #94-Ubuntu SMP Thu Aug 26 20:27:37 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux

### Important environment variables

```bash
DMS_DEBUG=1

POSTMASTER_ADDRESS=postmaster@MYDOMAIN.com

SSL_TYPE=manual
SSL_CERT_PATH=/etc/ssl/pki/mail.MYDOMAIN.com.crt
SSL_KEY_PATH=/etc/ssl/pki/mail.MYDOMAIN.com.key

ENABLE_LDAP=1
LDAP_START_TLS=yes
LDAP_SERVER_HOST=dc1.internal.MYDOMAIN.com
LDAP_SEARCH_BASE=dc=internal,dc=MYDOMAIN,dc=com
LDAP_BIND_DN=cn=Administrator,cn=users,dc=internal,dc=MYDOMAIN,dc=com
LDAP_BIND_PW=Ch@ngeMe
# Specifies how ldap should be asked for users.
LDAP_QUERY_FILTER_USER=(&(objectclass=person)(mail=%s))
# Specifies how ldap should be asked for groups.
LDAP_QUERY_FILTER_GROUP=(|)
# Specifies how ldap should be asked for aliases.
LDAP_QUERY_FILTER_ALIAS=(|)
# Specifies how ldap should be asked for domains.
LDAP_QUERY_FILTER_DOMAIN=(mail=*@%s)

DOVECOT_TLS=yes
DOVECOT_USER_FILTER=(&(objectclass=person)(sAMAccountName=%n))
DOVECOT_USER_ATTRS==uid=%{ldap:uidNumber},=gid=5000,=home=/var/mail/%Ln,=mail=maildir:~/Maildir
DOVECOT_PASS_ATTRS=sAMAccountName=user,userPassword=password
DOVECOT_AUTH_BIND=yes

ENABLE_SASLAUTHD=1
SASLAUTHD_MECHANISMS=ldap
SASLAUTHD_LDAP_SERVER=dc1.internal.MYDOMAIN.com
SASLAUTHD_LDAP_BIND_DN=cn=Administrator,cn=users,dc=internal,dc=MYDOMAIN,dc=com
SASLAUTHD_LDAP_SEARCH_BASE=dc=internal,dc=MYDOMAIN,dc=com
SASLAUTHD_LDAP_FILTER=(&(sAMAccountName=%U)(objectClass=person))
SASLAUTHD_LDAP_START_TLS=yes
SASLAUTHD_LDAP_TLS_CHECK_PEER=yes
SASLAUTHD_LDAP_TLS_CACERT_FILE=/etc/ssl/self-signed-pki/pki/ca.crt
```

### Relevant log output

```shell
Oct  3 18:17:40 mail postfix/master[2360]: daemon started -- version 3.4.14, configuration /etc/postfix
Oct  3 18:18:01 mail dovecot: imap-login: Login: user=<Administrator>, method=PLAIN, rip=85.127.7.5, lip=10.8.0.4, mpid=2396, TLS, session=<EFvwI3XNVttVfwe/>
Oct  3 18:18:01 mail dovecot: imap(Administrator)<2396><EFvwI3XNVttVfwe/>: Error: chdir(/var/mail/administrator/) failed: Permission denied (euid=10001(<unknown>) egid=5000(docker) missing +x perm: /var/mail/administrator, dir owned by 5000:5000 mode=0700)

Oct  3 18:18:01 mail dovecot: imap(Administrator)<2396><EFvwI3XNVttVfwe/>: Error: stat(/var/mail/administrator/Maildir/tmp) failed: Permission denied (euid=10001(<unknown>) egid=5000(docker) missing +x perm: /var/mail/administrator, dir owned by 5000:5000 mode=0700)
Oct  3 18:18:01 mail dovecot: imap(Administrator)<2396><EFvwI3XNVttVfwe/>: Error: stat(/var/mail/administrator/Maildir/.Sent/tmp) failed: Permission denied (euid=10001(<unknown>) egid=5000(docker) missing +x perm: /var/mail/administrator, dir owned by 5000:5000 mode=0700)

Oct  3 18:18:01 mail dovecot: imap(Administrator)<2396><EFvwI3XNVttVfwe/>: Error: stat(/var/mail/administrator/Maildir) failed: Permission denied (euid=10001(<unknown>) egid=5000(docker) missing +x perm: /var/mail/administrator, dir owned by 5000:5000 mode=0700)
Oct  3 18:18:01 mail dovecot: imap(Administrator)<2396><EFvwI3XNVttVfwe/>: Error: mkdir(/var/mail/administrator/Maildir) failed: Permission denied (euid=10001(<unknown>) egid=5000(docker) missing +x perm: /var/mail/administrator, dir owned by 5000:5000 mode=0700)
Oct  3 18:18:01 mail dovecot: imap(Administrator)<2396><EFvwI3XNVttVfwe/>: Error: stat(/var/mail/administrator/Maildir/.Sent/tmp) failed: Permission denied (euid=10001(<unknown>) egid=5000(docker) missing +x perm: /var/mail/administrator, dir owned by 5000:5000 mode=0700)
Oct  3 18:18:01 mail dovecot: imap(Administrator)<2396><EFvwI3XNVttVfwe/>: Error: stat(/var/mail/administrator/Maildir/tmp) failed: Permission denied (euid=10001(<unknown>) egid=5000(docker) missing +x perm: /var/mail/administrator, dir owned by 5000:5000 mode=0700)
Oct  3 18:19:01 mail dovecot: imap-login: Login: user=<Administrator>, method=PLAIN, rip=85.127.7.5, lip=10.8.0.4, mpid=2428, TLS, session=<QJuCJ3XNndtVfwe/>
Oct  3 18:19:01 mail dovecot: imap(Administrator)<2428><QJuCJ3XNndtVfwe/>: Error: chdir(/var/mail/administrator/) failed: Permission denied (euid=10001(<unknown>) egid=5000(docker) missing +x perm: /var/mail/administrator, dir owned by 5000:5000 mode=0700)
Oct  3 18:19:01 mail dovecot: imap(Administrator)<2428><QJuCJ3XNndtVfwe/>: Error: stat(/var/mail/administrator/Maildir) failed: Permission denied (euid=10001(<unknown>) egid=5000(docker) missing +x perm: /var/mail/administrator, dir owned by 5000:5000 mode=0700)
Oct  3 18:19:01 mail dovecot: imap(Administrator)<2428><QJuCJ3XNndtVfwe/>: Error: mkdir(/var/mail/administrator/Maildir) failed: Permission denied (euid=10001(<unknown>) egid=5000(docker) missing +x perm: /var/mail/administrator, dir owned by 5000:5000 mode=0700)
Oct  3 18:19:01 mail dovecot: imap(Administrator)<2428><QJuCJ3XNndtVfwe/>: Error: stat(/var/mail/administrator/Maildir/tmp) failed: Permission denied (euid=10001(<unknown>) egid=5000(docker) missing +x perm: /var/mail/administrator, dir owned by 5000:5000 mode=0700)
Oct  3 18:19:01 mail dovecot: imap(Administrator)<2428><QJuCJ3XNndtVfwe/>: Error: stat(/var/mail/administrator/Maildir/.Sent/tmp) failed: Permission denied (euid=10001(<unknown>) egid=5000(docker) missing +x perm: /var/mail/administrator, dir owned by 5000:5000 mode=0700)
Oct  3 18:19:01 mail dovecot: imap(Administrator)<2428><QJuCJ3XNndtVfwe/>: Error: mkdir(/var/mail/administrator/Maildir) failed: Permission denied (euid=10001(<unknown>) egid=5000(docker) missing +x perm: /var/mail/administrator, dir owned by 5000:5000 mode=0700)
Oct  3 18:19:01 mail dovecot: imap(Administrator)<2428><QJuCJ3XNndtVfwe/>: Error: stat(/var/mail/administrator/Maildir/.Sent/tmp) failed: Permission denied (euid=10001(<unknown>) egid=5000(docker) missing +x perm: /var/mail/administrator, dir owned by 5000:5000 mode=0700)
Oct  3 18:19:01 mail dovecot: imap(Administrator)<2396><EFvwI3XNVttVfwe/>: Connection closed (SELECT finished 59.933 secs ago) in=129 out=1026 deleted=0 expunged=0 trashed=0 hdr_count=0 hdr_bytes=0 body_count=0 body_bytes=0
Oct  3 18:19:01 mail dovecot: imap(Administrator)<2428><QJuCJ3XNndtVfwe/>: Error: stat(/var/mail/administrator/Maildir/tmp) failed: Permission denied (euid=10001(<unknown>) egid=5000(docker) missing +x perm: /var/mail/administrator, dir owned by 5000:5000 mode=0700)
```

### Other relevant information

```Markdown
1. Find the **Docker-Compose** file below:

version: '3.9'
services:
  mailserver:
    container_name: mailserver
    image: docker.io/mailserver/docker-mailserver:edge # TODO change the TAG to 10.2.0 when released.
    hostname: mail
    domainname: ${PUBLIC_DOMAIN_NAME}
    # dns: 1.1.1.1
    # To avoid conflicts with yaml base-60 float, DO NOT remove the quotation marks.
    # More information about the mailserver ports:
    # https://docker-mailserver.github.io/docker-mailserver/edge/config/security/understanding-the-ports/
    networks:
      - internal01
    ports:
      - "25:25"
      - "143:143"
      - "587:587"
      - "993:993"
    volumes:
      - mailserver-data:/var/mail/
      - mailserver-state:/var/mail-state/
      - mailserver-logs:/var/log/mail/
      - mailserver-config:/tmp/docker-mailserver/
      - /etc/localtime:/etc/localtime:ro
      - publicly-trusted-pki:/etc/ssl/pki/:ro
    restart: always
    stop_grace_period: 1m
    cap_add: [ "NET_ADMIN", "SYS_PTRACE" ]
    env_file:
      - mailserver/mailserver.env
    environment:
      # The following variables will be passed from the .env (if exist) to override the variables in mailserver.env.
      - DMS_DEBUG
      - POSTMASTER_ADDRESS
      - SSL_TYPE
      - SSL_CERT_PATH
      - SSL_KEY_PATH
      - ENABLE_LDAP
      - LDAP_START_TLS
      - LDAP_SERVER_HOST
      - LDAP_SEARCH_BASE
      - LDAP_BIND_DN
      - LDAP_BIND_PW
      - LDAP_QUERY_FILTER_USER
      - LDAP_QUERY_FILTER_GROUP
      - LDAP_QUERY_FILTER_ALIAS
      - LDAP_QUERY_FILTER_DOMAIN
      - DOVECOT_TLS
      - DOVECOT_USER_FILTER
      - DOVECOT_USER_ATTRS
      - DOVECOT_PASS_ATTRS
      - DOVECOT_AUTH_BIND
      - ENABLE_SASLAUTHD
      - SASLAUTHD_MECHANISMS
      - SASLAUTHD_LDAP_SERVER
      - SASLAUTHD_LDAP_BIND_DN
      - SASLAUTHD_LDAP_SEARCH_BASE
      - SASLAUTHD_LDAP_FILTER
      - SASLAUTHD_LDAP_START_TLS
      - SASLAUTHD_LDAP_TLS_CHECK_PEER
      - SASLAUTHD_LDAP_TLS_CACERT_FILE


volumes:
  mailserver-data:
  mailserver-state:
  mailserver-logs:
  mailserver-config:

internal01 network and the volume of the public key are defined in another docker-compose.yml file. But this should not be an issue.

2. Make sure the container does not validate the CA in the ldap.conf file, as is seems that there is an issue here (the certificate seems to be refused even if it is valid). In simple words, execute the following on the host and this will do the job:

    docker exec mailserver sh -c 'echo "TLS_REQCERT   never" > /etc/ldap/ldap.conf'

Then of course restart the container.
```

### What level of experience do you have with Docker and mail servers?

Trust me, I'm a (computer) engineer! [expert]

### Code of conduct

- [X] I have read this project's [Code of Conduct](https://github.com/docker-mailserver/docker-mailserver/blob/master/CODE_OF_CONDUCT.md) and I agree
- [X] I have read the [README](https://github.com/docker-mailserver/docker-mailserver/blob/master/README.md) and the [documentation](https://docker-mailserver.github.io/docker-mailserver/edge/) and I searched the [issue tracker](https://github.com/docker-mailserver/docker-mailserver/issues?q=is%3Aissue) but could not find a solution

**Comment by MohammedNoureldin** [comment]
Seems to be related to https://github.com/docker-mailserver/docker-mailserver/issues/475

**Comment by polarathene** [comment]
Just to clarify, this problem only occurs with LDAP setup? And you're triggering it via an LDAP account doing IMAP login to Dovecot?

We lack a maintainer with LDAP experience AFAIK, so ideally we want to have a test case added to reproduce the problem upfront (_so once it's fixed, if there is ever a regression we can catch it again_).

> The suggested solution there to set both uid (and the gid of course) 5000 is not a optimal.

Is both required?

`uid` gives you permissions of the owner of the file/directory (_eg docker/5000_), while `gid` gives anyone in the separate docker group (_also 5000, but distinct from uid and docker user_) a reduced set of permissions, but these can be more than users not belonging to the group.

---

Regarding the cert setup.

> Make sure the container does not validate the CA in the `ldap.conf` file, as is seems that there is an issue here (the certificate seems to be refused even if it is valid).

I see that you are using `SSL_TYPE=manual`, are you using your own internal/private CA? If so the trust chain is missing within the container. I still need to find time to refactor the code here, in the meantime you can try [`SSL_TYPE=custom`](https://docker-mailserver.github.io/docker-mailserver/edge/config/environment/#ssl_type) which is an existing workaround.

That requires combining your PEM files in this order: `private-key + certificate + CA chain`, if there is no intermediate CA, the CA chain part would just be the cert for your CA. Your `.crt` and `.key` extensions should be PEM encoded (usually are), if so they're text files you can open and see some lines like `--- BEGIN <something> ---`; just copy/paste each files content in the mentioned order to a new file and it should be correct.

As this is an undocumented workaround, viewing the relevant code is somewhat helpful:

https://github.com/docker-mailserver/docker-mailserver/blob/cd7677b6f020099a85ae35739cdd7a485491bc92/target/scripts/startup/setup-stack.sh#L982-L1001

This looks for `/tmp/docker-mailserver/<hostname>-full.pem`, so in your local `docker-data/dms/config/` folder, place in your `mail.example.com-full.pem` file and set `SSL_TYPE=custom`.

**EDIT:** Nope nevermind, [these test docs I wrote](https://github.com/docker-mailserver/docker-mailserver/tree/master/test/test-files/ssl/example.test) state at the bottom not to add the root CA to the end of the `<hostname>-full.pem` file.. [self-signed](https://github.com/docker-mailserver/docker-mailserver/blob/cd7677b6f020099a85ae35739cdd7a485491bc92/target/scripts/startup/setup-stack.sh#L1054-L1097) may work. It expects `<hostname>-cert.pem` and `<hostname>-key.pem` files at the same `config/` location, these would just be renaming your `.crt` and `.key` files, in addition the CA cert is provided at the same location in a subfolder:  `demoCA/ca-cert.pem`. This may only work with Postfix software, but if it resolves the invalid error let me know :)

This assumes you're not using a CA that should already be publicly trusted. If you are let me know as that should work fine with `SSL_TYPE=manual`.

---

> `docker exec mailserver sh -c 'echo "TLS_REQCERT   never" > /etc/ldap/ldap.conf'`

I think you can use `LDAP_TLS_REQCERT=never` as an ENV to achieve the same.

**Comment by MohammedNoureldin** [comment]
Hi @polarathene,

Yes it is triggered when I click refresh on my mail client (namely IMAP login to Dovecot).

I cannot tell exactly if that solution with 5000 works, but from what I see it should. I am still doing some tests, I will let you know as soon as I know.

Regarding the certificate, I am using a full-chain generated by lets encrypt by another docker container (so a combination of the certificate itself and the intermediate and CA certs used to sign the cert). I believe the certificate should not be an issue here.

> I think you can use LDAP_TLS_REQCERT=never as an ENV to achieve the same.

Oh I will give it a try thanks!

**Comment by MohammedNoureldin** [comment]
Just a quick feed back regarding

> DAP_TLS_REQCERT=never

This seems to have to effect. At least it cannot replace the system ldap.conf

**Comment by MohammedNoureldin** [comment]
I am not able to fix it by using a static uid=5000, but I am still trying

**Comment by polarathene** [comment]
> I believe the certificate should not be an issue here.

As LetsEncrypt is a public CA I agree. I am not sure why it would output an error about not being valid then.. Do you know specifically what service is complaining?

> This seems to have no effect. At least it cannot replace the system ldap.conf

Ah.. we might want to provide some way to support configuring that, but first it's probably helpful to know why the cert is not considered valid :sweat_smile: 

> I am not able to fix it by using a static uid=5000, but I am still trying

If all we really need to do is add `+x` permissions and that's not due to something LDAP specific, then @casperklein or @georglauterbach probably knows if the related startup script that does `chown -R` to `/var/mail` is a valid fix.

Other than that, I don't think I can be of much help here until I find time to get familiar with LDAP. We have [an LDAP test](https://github.com/docker-mailserver/docker-mailserver/blob/master/test/mail_with_ldap.bats) which is mostly a shell script with some test functions. It uses a docker image to provide the LDAP container to pair with `docker-mailserver` and set some ENV.

I assume that can reproduce the problem, or could be adjusted to do so. I don't presently have time to allocate.

---

In the meantime, since this is presumably LDAP specific which is not as well supported AFAIK, and the other maintainers may also be limited in time they can spare; you may want to try workaround the issue with [`user-patches.sh`](https://docker-mailserver.github.io/docker-mailserver/edge/config/advanced/override-defaults/user-patches/).

That allows you to run a shell script after `docker-mailserver` completes container startup. You should be able to use this to run a command to set `+x` on `/var/mail`. You have [a few options to approach that](https://stackoverflow.com/a/13378069), but `chmod -R +x /var/mail` should be sufficient.

**Comment by georglauterbach** [comment]
We're potentially missing fine-grained permission here. I will have a look later.

This issue should however not affect the release of `v10.2.0` as it is most likely not a regression.

**Comment by georglauterbach** [comment]
@MohammedNoureldin so executing

``` BASH
chown -R 10001:10001 /var/mail/administrator
```

should do the trick in this case, right?

---

I can image doing the following: Default permission level is `drwxr-xr-x` (`755`). Running

``` BASH
find /var/mail/ -maxdepth 1 -type d -not -iname "*\.*" -exec chmod -R 755 {} \;
```

which _should_ set the correct permissions. @MohammedNoureldin can you try that?

---

In addition, why does `/var/mail` and `/var/mail-state` have `777` permissions? Can someone tell me @docker-mailserver/maintainers? This seems too open to me and not needed. Shouldn't `755` suffice too?

**Comment by casperklein** [comment]
> In addition, why does /var/mail and /var/mail-state have 777 permissions?

Why do you think it is this way? I cannot confirm this. My files to not have 777 permission.

**Comment by MohammedNoureldin** [comment]
I am trying now with 10.1.2, because I got it somehow last week with some manipulations working. The problem this functional status is not reproduceable now :( (I will keep trying). As soon as I got a fully functional with 10.1.2, I will try with 10.2.0.

At the moment, I am using */home/mail/%Ln* as home directory. And I am giving it manually the permission: **2775**. This helps to get rid of the permission issue. However, I am still not able to get a fully functional email server, where receiving emails seems to work somehow, I am still not able to refresh the folders or send emails, but as said, I am still working on it.

**Comment by georglauterbach** [comment]
> > In addition, why does /var/mail and /var/mail-state have 777 permissions?
> 
> Why do you think it is this way? I cannot confirm this. My files to not have 777 permission.

Then this a K8s thing and my fault accordingly :D
> 
> > will create some problems too and should be "fixed" accordingly.
> 
> What is problematic with that?

This would again overwrite the owners of all files in `var/mail` which we want to avoid...

**Comment by casperklein** [comment]
fyi: in my setup, no execute permission is missing:

```
root@mail:/var/mail# ll
total 20K
drwxr-xr-x 4 docker docker 4.0K May 29 00:55 domain1
drwxr-xr-x 3 docker docker 4.0K Mar  4  2020 domain2
drwxr-xr-x 3 docker docker 4.0K Mar  4  2020 domain3
```

I am pretty sure, this is something LDAP related. In a non LDAP setup, the permissions are correct.

**Comment by MohammedNoureldin** [comment]
Please stop investigating this issue in the meanwhile, I am not sure how and why, but the error is not reproduceable for me at the moment. Let us keep this issue open for the next few days until I know why the issue does not happen anymore (of course even after removing and recreating the container and cleaning the dependent volumes). I need to check what I changed in my docker-compose file.

Anyway, even before the error being not reproduceable anymore, I can confirm that the correct permissions for the ***mail*** folder should be 2770 at least (at the moment it is 2775, which should not harm). The owner at the moment is 5000:5000 which is totally fine.

Can I just ask ask a side-question, what `PERMIT_DOCKER` exactly does? In case of LDAP server (don't know if LDAP can change it), what value should it take? Somehow I don't understand its documentation and what it exactly does. I just did not want to open a new issue as a question for this, but if you think it is needed, I will. Thank you!

**Comment by georglauterbach** [comment]
`PERMIT_DOCKER` is a somewhat misleading name (was before my time). It's about network settings, which is especially important for non-Docker/Docker Compose setups. This is most-likely not LDAP related.

**Comment by MohammedNoureldin** [comment]
> `PERMIT_DOCKER` is a somewhat misleading name (was before my time). It's about network settings, which is especially important for non-Docker/Docker Compose setups. This is most-likely not LDAP related.

Thank you @georglauterbach,

Could you pleae just a little bitore about it? With a use case example?

In one of ldap examples you used host, and in another you did not use docker permit at all, but I still don't understand when I need it.

**Comment by MohammedNoureldin** [comment]
> I don't remember when I used `PERMIT_DOCKER=1` in an example, but if I did, it was by accident. It's basically whether the containers runs behinder Docker's NAT or whether network devices are directly mapped inside the container. The `mailserver.env` description explains a bit more.

Here:

https://docker-mailserver.github.io/docker-mailserver/edge/config/advanced/auth-ldap/#ldap-setup-examples

**Comment by georglauterbach** [comment]
I really didn't use it myself, so my experience is rather limited :) Usually, I've only ever seen this variable being used in more complicated setups, and if you haven't thought about networking because your setup may require it, you are most likely not in the need to use the variable:)

**Comment by MohammedNoureldin** [comment]
Well, let us come back to this issue. Somehow it is reproducable somehow. Maybe also because I restarted the container (not recreated it)? I am not sure what I have done, but the error is here again.

```
Oct 18 19:17:02 mail dovecot: imap-login: Login: user=<Administrator>, method=PLAIN, rip=85.127.7.191, lip=10.8.0.2, mpid=2186, TLS, session=<vj+5Y6XO6NZVfwe/>
Oct 18 19:17:02 mail dovecot: imap(Administrator)<2186><vj+5Y6XO6NZVfwe/>: Error: chdir(/var/mail/administrator/) failed: Permission denied (euid=10000(<unknown>) egid=5000(docker) missing +x perm: /var/mail/administrator, dir owned by 5000:5000 mode=0700)
Oct 18 19:17:02 mail dovecot: imap(Administrator)<2186><vj+5Y6XO6NZVfwe/>: Error: stat(/var/mail/administrator/Maildir) failed: Permission denied (euid=10000(<unknown>) egid=5000(docker) missing +x perm: /var/mail/administrator, dir owned by 5000:5000 mode=0700)
Oct 18 19:17:02 mail dovecot: imap(Administrator)<2186><vj+5Y6XO6NZVfwe/>: Error: mkdir(/var/mail/administrator/Maildir) failed: Permission denied (euid=10000(<unknown>) egid=5000(docker) missing +x perm: /var/mail/administrator, dir owned by 5000:5000 mode=0700)
Oct 18 19:17:02 mail dovecot: imap(Administrator)<2186><vj+5Y6XO6NZVfwe/>: Error: stat(/var/mail/administrator/Maildir/tmp) failed: Permission denied (euid=10000(<unknown>) egid=5000(docker) missing +x perm: /var/mail/administrator, dir owned by 5000:5000 mode=0700)
Oct 18 19:17:02 mail dovecot: imap(Administrator)<2186><vj+5Y6XO6NZVfwe/>: Error: stat(/var/mail/administrator/Maildir/.Sent/tmp) failed: Permission denied (euid=10000(<unknown>) egid=5000(docker) missing +x perm: /var/mail/administrator, dir owned by 5000:5000 mode=0700)
Oct 18 19:17:02 mail dovecot: imap(Administrator)<2186><vj+5Y6XO6NZVfwe/>: Error: mkdir(/var/mail/administrator/Maildir) failed: Permission denied (euid=10000(<unknown>) egid=5000(docker) missing +x perm: /var/mail/administrator, dir owned by 5000:5000 mode=0700)
Oct 18 19:17:02 mail dovecot: imap(Administrator)<2186><vj+5Y6XO6NZVfwe/>: Error: stat(/var/mail/administrator/Maildir/.Sent/tmp) failed: Permission denied (euid=10000(<unknown>) egid=5000(docker) missing +x perm: /var/mail/administrator, dir owned by 5000:5000 mode=0700)
Oct 18 19:17:02 mail dovecot: imap(Administrator)<2186><vj+5Y6XO6NZVfwe/>: Connection closed (SELECT finished 0.048 secs ago) in=139 out=1263 deleted=0 expunged=0 trashed=0 hdr_count=0 hdr_bytes=0 body_count=0 body_bytes=0
Oct 18 19:17:48 mail postfix/anvil[2054]: statistics: max connection rate 1/60s for (smtps:85.127.7.191) at Oct 18 19:13:09
Oct 18 19:17:48 mail postfix/anvil[2054]: statistics: max connection count 1 for (smtps:85.127.7.191) at Oct 18 19:13:09
Oct 18 19:17:48 mail postfix/anvil[2054]: statistics: max cache size 1 at Oct 18 19:13:09
Oct 18 19:18:10 mail postfix/smtpd[2062]: timeout after END-OF-MESSAGE from localhost[127.0.0.1]
Oct 18 19:18:10 mail postfix/smtpd[2062]: disconnect from localhost[127.0.0.1] ehlo=1 mail=1 rcpt=1 data=1 commands=4
Oct 18 19:19:28 mail postfix/smtpd[2107]: timeout after END-OF-MESSAGE from localhost[127.0.0.1]
Oct 18 19:19:28 mail postfix/smtpd[2107]: disconnect from localhost[127.0.0.1] ehlo=1 mail=1 rcpt=1 data=1 commands=4
Oct 18 19:22:14 mail postfix/qmgr[2010]: 900D246264: from=<m.i.noureldin@gmail.com>, size=4565, nrcpt=1 (queue active)
Oct 18 19:22:14 mail dovecot: lmtp(2315): Connect from local
Oct 18 19:22:15 mail dovecot: lmtp(administrator@DOMAIN.com)<2315><6+OOMGbJbWELCQAAUGJTbw>: Error: sieve: file storage: Failed to stat sieve storage path: stat(/var/mail/administrator/sieve) failed: Permission denied (euid=10000(<unknown>) egid=5000(docker) missing +x perm: /var/mail/administrator, dir owned by 5000:5000 mode=0700)
Oct 18 19:22:15 mail dovecot: lmtp(administrator@DOMAIN.com)<2315><6+OOMGbJbWELCQAAUGJTbw>: Error: sieve: Failed to access user's personal storage
Oct 18 19:22:15 mail dovecot: lmtp(administrator@DOMAIN.com)<2315><6+OOMGbJbWELCQAAUGJTbw>: Error: stat(/var/mail/administrator/Maildir/tmp) failed: Permission denied (euid=10000(<unknown>) egid=5000(docker) missing +x perm: /var/mail/administrator, dir owned by 5000:5000 mode=0700)
Oct 18 19:22:15 mail dovecot: lmtp(administrator@DOMAIN.com)<2315><6+OOMGbJbWELCQAAUGJTbw>: msgid=<CA+hEKsNutUW_tw3tjmtviCeoVuiSKdvLvn_-KpApaCW3G47s_w@mail.gmail.com>: save failed to open mailbox INBOX: stat(/var/mail/administrator/Maildir/tmp) failed: Permission denied (euid=10000(<unknown>) egid=5000(docker) missing +x perm: /var/mail/administrator, dir owned by 5000:5000 mode=0700)
Oct 18 19:22:15 mail dovecot: lmtp(2315): Disconnect from local: Client has quit the connection (state=READY)
Oct 18 19:22:15 mail postfix/lmtp[2314]: 900D246264: to=<administrator@DOMAIN.com>, relay=mail.DOMAIN.com[/var/run/dovecot/lmtp], delay=467, delays=466/0.16/0.04/0.36, dsn=4.2.0, status=deferred (host mail.DOMAIN.com[/var/run/dovecot/lmtp] said: 451 4.2.0 <administrator@DOMAIN.com> Internal error occurred. Refer to server log for more information. [2021-10-18 19:22:15] (in reply to end of DATA command))
```

I am still trying to gather some more helpful information.

**Comment by georglauterbach** [comment]
Most likely die to a problematic start after all. With a clean state, all _should_ be better now. Keep us updated:)

**Comment by MohammedNoureldin** [comment]
IMPORTANT UPDATE:

It is **restarting** the container! Restarting the container ruins the owners of folders in */var/mail*

Could anyone try to fix it please? This is really a killer.

**Comment by polarathene** [comment]
> Somehow the owner of the whole user directory in mail folder was changed to 5000 (it was 10000 yesterday, which is the correct one I defined in my LDAP). 5000 as a group owner is fine, the user 5000 is wrong.

Is user 10000 a member of the group 5000? If so, isn't the issue more due to group permissions?

I don't know the reasoning behind the chown in `fixes-stack.sh`, if you change it to 10000 for UID, then you may be fixing one problem while introducing another, assuming the UID 5000 requires to interact with the file content in a similar manner. If it doesn't, then you could look into contextually changing UID/GID used when LDAP is in use...?

Is there any reason the LDAP UID cannot cahnge from 10000 to 5000?

**Comment by MohammedNoureldin** [comment]
https://github.com/docker-mailserver/docker-mailserver/pull/2256

**Comment by MohammedNoureldin** [comment]
@polarathene I have good experience with LDAP, for me the Docs of LDAP are fine, I read the whole page and it looks fine. I am pretty sure here that the solution is just to not change the permissions recursively.

For me this makes totally sense that every user has only permissions on his own folder. Using a hard coded value in LDAP config to use uid=5000 is not the optimal solution I would say, IMHO.

**Comment by MohammedNoureldin** [comment]
Hey @polarathene, I would not put this on low priority. This corrupts the whole permissions for LDAP users, which is really a bad thing. So if not high, this is at least a medium priority issue, IMHO.

It is gonna take a bit till I check the tests, because I have some circumstances where I almost don't sit on the computer. I saw that someone contributed in the pull request and I hoped that he could finish it, but he did not respond to my request. I will poke him and ask if he can support in this. Till that I will try as soon as possible to fix the tests.

**Comment by Encephala** [comment]
Just to chime in, I'm getting the same issue. Directories in `/var/mail/` are being made for LDAP users upon authenticating to IMAP, and the directory UID and GID are appropriate, `14605:14605` for my personal account. However, upon restarting the container, the directory owner is recursively reset to `vmail:vmail`. 
Seems to be a dealbreaker for our usecase.

Well, The fix to hardcode dovecot's `mail_uid` and `mail_gid` to 5000 (as described [here](https://github.com/docker-mailserver/docker-mailserver/issues/475#issuecomment-899652084)) does fix our problem.

I can see and respect your argument on limited dev time and other priorities, but then I'd argue that "LDAP auth" should not be communicated in the README.

**Comment by MohammedNoureldin** [comment]
Hi @Encephala 

Do mean the solution I proposed? I fixed it as shown. However, I did not have enough time to check the unit tests. That is why my fiy did not get merged. Do you have time to fix them? It would be great if we can fix it together.

**Comment by Encephala** [comment]
Sorry for the late response, summer holidays got in the way.

I've been looking over the PR and the many comments (wow you all have put a lot of effort into this), but I'm also losing the forest for the trees.

A thought I had: perhaps we should use `chown` recursively but only going one directory deep, i.e.

`find /var/mail -maxdepth 1 -exec chown 5000:5000 {} \;`

If I recall correctly, in an LDAP setup the dir `/var/mail/domain.tld` is still owned by user `vmail` (UID 5000), but all directories inside of there are owned by the respective users (with local users created through the setup script, I can imagine the tests are failing because the `.../domain.tld` directory isn't writable for Dovecot, but I think with appropriate permissions on that directory set (i.e. traversable for `other`) LDAP might still work this way.

**Comment by polarathene** [comment]
> A thought I had: perhaps we should use `chown` recursively but only going one directory deep, i.e.
> 
> `find /var/mail -maxdepth 1 -exec chown 5000:5000 {} \;`

I've been getting familiar and improving on the LDAP support recently. The current `chown` logic lives here:

https://github.com/docker-mailserver/docker-mailserver/blob/ad8b618b461236fe504cdbb0d0ddf0bff0283840/target/scripts/helpers/utils.sh#L34-L46

I have a PR related to mail encryption to revive at some point, where I think it also matters that the `chown` works correctly, although that probably won't be correct for LDAP.

> If I recall correctly, in an LDAP setup the dir `/var/mail/domain.tld` is still owned by user `vmail` (UID 5000), but all directories inside of there are owned by the respective users.

When Dovecot is writing content to the mail storage, it is root but will use the ownership it's told to with `uid` / `gid` this is globally the `docker` user and group, both assigned `5000`. LDAP users affected here are pulling an override based on the associated users LDAP attributes for UID + GID.

The LDAP setup may have also chosen to configure the mail storage differently, especially with v13 where the mail home directory for non-LDAP has been moved into a subdir of the `/var/mail/domain/user/` folder. Some may IIRC choose `/var/mail/user` if they didn't feel a need for a distinct domain 🤷‍♂️ 

---

The issue isn't so much LDAP, as it is straying from our defaults that DMS expects 😅 

I think that the `docker` user/group may also be effectively only for Dovecot, so it's more like `dovecot` rather than `docker` as a user/group (_or as you noted, locally is understood as `vmail`?_). Pretty sure `/var/mail` had to be adjusted during startup from the original defaults to what DMS wanted.

> I can see and respect your argument on limited dev time and other priorities, but then I'd argue that "LDAP auth" should not be communicated in the README.

LDAP auth should still work. It's the mail storage / delivery that's not. Same would happen if someone changed from our defaults for Dovecot UserDB without LDAP.

I'm open to skipping the `chown` when using LDAP so long as tests pass. My understanding is the `chown` is meant to fix:
- `/var/mail` has different ownership during startup, volume mount or not.
- During startup or change event, the default account provisioner (file-based) will create any additional user directories in `/var/mail`, but the scripts do so as `root:root`. They could fix ownership then and there but have been relying on the `chown -R` helper.
- A feature PR I need to revive creates some files, but also needs to block on waiting until the `check-for-changes.sh` / Dovecot logic for accounts has done it's thing. Feature may not be compatible with LDAP either way, so non-concern.

If an image build with that helper function disabled works well for LDAP, and that change doesn't fail any of our tests (_they can be adjusted when it makes sense to do so_), accepting the change should be ok. If `/var/mail` does need to be owned by `5000:5000`, even for LDAP that can be addressed. Not sure about the domain folder, Dovecot should be able to create that I think? (_either based on ownership of `/var/mail`, or the UID+GID at the time for the user, but you don't want the user attributes affecting the domain directory_).

**Comment by andrew-techwellington** [comment]
this is quite sad its been a bug since 2017 and still marked as low priority. It basically makes LDAP unusable unless you want to make /var/mail/ world writable every time the container starts. Are there any plans to fix this or is LDAP support going away?

**Comment by MohammedNoureldin** [comment]
@andrew-techwellington I proposed a fix long time ago, it fixed the issue but the unit tests failed. If you have time, could you check it and adapt the Unit Test? @Encephala also wanted to check it, I am not sure if get got time for this.

If you can check why the unit tests did not work, I can also support if you need to know any details related to LDAP, as far as I can.

**Comment by georglauterbach** [comment]
@andrew-techwellington we have lots of people complaining about this; what we need though, are people forking, adjusting and merging new code. This may sound harsh, but @MohammedNoureldin is correct in pointing out that we have a MR, but we need more help. That's how FOSS works..

**Comment by MohammedNoureldin** [comment]
Here is the PR, @andrew-techwellington @Encephala  in case you have some time to investigate why the tests failed. I am not sure if anyone else provided another PRs since I created my PR.

**Comment by andrew-techwellington** [comment]
Thanks for the replies guys, and I realize I came across rudely so I apologize for that. I did some more research and I think my issue might be because I am running DMS on a synology NAS and using the synology LDAP server. I did a complete wipe of the data files and rebuilt the container and volumes and the error stopped appearing. I don't know if that means its fixed forever, but I did see this is a known issue with synology DSM. I am limited with my linux knowledge, I am more a powershell guy than bash but I will check out the unit test you posted.

**Comment by polarathene** [comment]
> I am limited with my linux knowledge

- The issue AFAIK is related to proper Linux file/dir permissions management of `/var/mail`, which mounting your own volume to strips away.
- We shouldn't need to enforce recursively deep static change, that can be deferred to a `setup ...` CLI helper.
- I have a modern issue for tracking / discussing a way forward (_that you've already chimed in_) here: https://github.com/docker-mailserver/docker-mailserver/issues/3557#issuecomment-1741708401
- I will get around to it when I can, I only have so much volunteer time available and I try to balance that (_sometimes it's biased to effort required / priorities_).
  - Presently I actively participate in triage / troubleshooting on DMS and PR reviews.
  - My current DMS task is to rewrite documentation for the LDAP refactoring I've been doing, and this is a very tedious task that I also set a high bar for myself to write well for users. As such I do get bored of it and go on detours elsewhere, documentation tasks slow my contributions down quite a bit.
  - Once that's resolved I have some stale PRs to revive, and after that I will wrap up the test suite revision.
  - The issue here itself is researched and understood fairly well, I'd just like more confidence from tests before bundling it with all the other bigger changes in v13. In the meantime, feel free to share confirmation on what changes work for you.
